sound/soc/sof/core.c

```diff
@@ -325,6 +325,7 @@ int snd_sof_device_probe(struct device *dev, struct snd_sof_pdata *plat_data)
 	spin_lock_init(&sdev->ipc_lock);
 	spin_lock_init(&sdev->hw_lock);
 	mutex_init(&sdev->power_state_access);
+	mutex_init(&sdev->cores_status_mutex);
```

Comment:
thinking a bit more, do we really need this mutex? In reality the core power up/power down is already protected by the IPC. So all that this mutex would do is protect the core_ref count which doesnt need to be protected I think.

Reply:
this needs to be analyzed a bit more, adding a mutex just based on hypotheses is not good enough in my book, you need to have a clear use model where cores might be enabled/disabled concurrently.

Doing this will actually add a lot of clarity on test coverage, etc.
Please revisit this, thanks.

Reply:
> this needs to be analyzed a bit more, adding a mutex just based on hypotheses is not good enough in my book, you need to have a clear use model where cores might be enabled/disabled concurrently.
> 
> Doing this will actually add a lot of clarity on test coverage, etc.
> Please revisit this, thanks.


@plbossart It's actually not hypotheses, this happens everyday during our CI multiple-pipeline-xx.sh, e.g.
```
1: terminal one, 'aplay -Dhw:0,0': 
    Mutex hold [1]
    1.a ask for power on core 1 -> 
        1.aa set driver registers to power up core 1
        1.ab send IPC to FW to enable core 1.
    1.b update the enabled_cores_mask.
    Mutex release [1]

2: terminal two, 'aplay -Dhw:0,1': 
    Mutex hold [2]
    2.a ask for power on core 1 -> 
        2.aa set driver registers to power up core 1
        2.ab send IPC to FW to enable core 1.
    2.b update the enabled_cores_mask.
    Mutex release [2]
```
The 2 'aplay's are 2 separated thread, if we don't have mutex hold in [1] and [2], we could hit kinds of errors like the driver registers are set twice, or the core enable IPC will be sent twice, or the enabled_cores_mask is inconsistent, ...

Reply:
Sorry @, I don't fully buy the argument.

You have to power-up cores at the .prepare level, and this has to be done via the pipeline widget and hence DAPM.
I would think that DAPM has some sort of consistency already.